server/src/main/java/org/apache/iotdb/db/engine/compaction/task/CompactionRecoverTask.java

```diff
@@ -20,102 +20,481 @@
+        LOGGER.error(
+            "{} [Compaction][Recover] fail to delete target file {}, this may cause data incorrectness",
+            fullStorageGroupName,
+            compactionModFile);
+        handleSuccess = false;
+      }
     }
+    return handleSuccess;
+  }
+
+  /**
+   * This method find the File object of given filePath by searching it in every data directory. If
+   * the file is not found, it will return null.
+   */
+  private File getFileFromDataDirs(String filePath) {
+    String[] dataDirs = IoTDBDescriptor.getInstance().getConfig().getDataDirs();
+    for (String dataDir : dataDirs) {
+      File f = new File(dataDir, filePath);
+      if (f.exists()) {
+        return f;
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public boolean equalsOtherTask(AbstractCompactionTask otherTask) {
+    if (otherTask instanceof CompactionRecoverTask) {
+      return compactionLogFile.equals(((CompactionRecoverTask) otherTask).compactionLogFile);
+    }
+    return false;
+  }
+
+  @Override
+  public boolean checkValidAndSetMerging() {
+    return compactionLogFile.exists();
   }
 
-  private void recoverCrossCompactionFromOldVersion() throws Exception {
-    // check whether there is old compaction log from previous version (<0.13)
-    File mergeLogFromOldVersion =
-        new File(
-            tsFileManager.getStorageGroupDir()
-                + File.separator
-                + CompactionLogger.CROSS_COMPACTION_LOG_NAME_FROM_OLD);
-    if (mergeLogFromOldVersion.exists()) {
-      logger.info("calling cross compaction task to recover from previous version.");
-      IoTDBDescriptor.getInstance()
-          .getConfig()
-          .getCrossCompactionStrategy()
-          .getCompactionRecoverTask(
-              logicalStorageGroupName,
-              virtualStorageGroupId,
-              0L,
-              mergeLogFromOldVersion,
-              tsFileManager)
-          .call();
+  @Override
+  public void setSourceFilesToCompactionCandidate() {
+    // do nothing
+  }
+
+  @Override
+  public void resetCompactionCandidateStatusForAllSourceFiles() {
+    // do nothing
+  }
+
+  /**
+   * Used to check whether it is recoverd from last version (<0.13) and perform corresponding
+   * process.
+   */
+  private class CompactionRecoverFromOld {
+
+    /** Return whether cross compaction log file is from previous version (<0.13). */
+    private boolean isOldCrossCompactionLog() {
+      return compactionLogFile
+          .getName()
+          .equals(CompactionLogger.CROSS_COMPACTION_LOG_NAME_FROM_OLD);
+    }
+
+    /** Return whether inner compaction log file is from previous version (<0.13). */
+    private boolean isOldInnerCompactionLog() {
+      return compactionLogFile.getName().startsWith(tsFileManager.getStorageGroupName());
+    }
+
+    /** Delete tmp target file and compaction mods file. */
+    private boolean handleCrossCompactionWithAllSourceFilesExistFromOld(
```

Comment:
The method should specify which versions could be handle.

Reply:
Resolved